---
title: "Scripts to create the main figures: Figure 1"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r load_library, echo=FALSE}
suppressPackageStartupMessages( library(tidyverse) )
suppressPackageStartupMessages( library(edgeR) )
suppressPackageStartupMessages( library(limma) )
suppressPackageStartupMessages( library(SummarizedExperiment) )
suppressPackageStartupMessages( library(qvalue) )
suppressPackageStartupMessages( library(EnhancedVolcano) )
suppressPackageStartupMessages( library(gridExtra) )
suppressPackageStartupMessages( library(ComplexHeatmap) )
suppressPackageStartupMessages( library(ggpubr) )
```

```{r setup, echo=FALSE}
knitr::opts_knit$set(root.dir = '~/Desktop/Workspace/HDBR/github-repo/sexDE_prenatal_brain/')
```

```{r param, echo=FALSE}
threshold_qvalue <- 0.01
```


# Figure 1 B

Boxplot of the pseudotime grouped by developmental stages and colored by sex.

## Load data

```{r figure1B_load_data, echo=FALSE}
# load pseudotime data
load("data/2-pseudotime/forebrain_pseudotime_analysis.RData")
```

## Prepare the data

Convert the developmental stages to a continuous variable.

```{r convert2continuous, echo=FALSE}
metadata$DevStageCont <- as.integer(metadata$DevStage)
```

Compute Kendall correlation between the pseudotime and the new continuous developmental stage.

```{r correlation_kendall, echo=FALSE}
corTestRes_Kendall <- cor.test(metadata$DevStageCont, metadata$pseudotime, method = "kendall")
corTestRes_Kendall
```

## Plot 

Scatter plot of the pseudotime and the continuous developmental stage.

```{r scatter_plot, echo=FALSE}
p <- ggplot(metadata, aes(DevStage, pseudotime))
p <- p + geom_boxplot(fill = "#DCDCDC", outlier.shape = NA)
p <- p + theme_minimal() + xlab("Developmental stages")
p <- p + geom_jitter(aes(color = Sex), position=position_jitter(0.2), show.legend = FALSE)
p <- p + annotate("text", x = 6.5, y = -0.7, label = sprintf("tau = %.2f\n p-value = %.2e", corTestRes_Kendall$estimate, corTestRes_Kendall$p.value))
p <- p + scale_color_manual(values = c("1" = "darkblue", "2" = "darkred"))
p <- p + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 10),
               axis.text.y = element_text(size = 10),
               axis.title.x = element_text(size = 14),
               axis.title.y = element_text(size = 14))
p
# save plot
ggsave("1B_boxplot_devStages_vs_pseudotime.svg", 
       path = "images/figures/",
       plot = p, width = 10, height = 8, units = "cm")
ggsave("1B_boxplot_devStages_vs_pseudotime.png", 
       path = "images/figures/",
       plot = p, width = 10, height = 8, units = "cm")
```


# Figure 1 C

Volcanoplot of the sex-DE genes colored by sex (without autosomes). 
And barplot of the number of sex-DE genes

## Load and prepare data

```{r figure1C_load_data, echo=FALSE}
# load sex-DE data
load("data/3-DE/normCounts_dream_pseudotime_sva_ind_interaction_phenotypes.RData")
de_results <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction.txt",
                         header = TRUE)
```

From the full list of sex-DE genes, remove XIST and the protein-coding genes on the Y-chromosome.

```{r remove_training_genes, echo=FALSE}
genes2remove <- c("XIST", as.character(de_results$gene_name[which(de_results$chr == "chrY" & de_results$gene_type == "protein_coding")]))
de_results_filtered <- de_results[which(!de_results$gene_name %in% genes2remove),]
```

Create a data frame containing only the results for the autosomes.

```{r autosomes, echo=FALSE}
chrNonAuto <- c("chrY", "chrX", "chrM")
de_results_autosomes <- de_results_filtered[which(!de_results_filtered$chr %in% chrNonAuto),]
```

## Barplot

```{r function_barplot, echo=FALSE}
barplot <- function(data, legend = TRUE, nrowLegend = 4,
                    colors = c("Male-autosomes" = 'darkblue', "Male-sex chr." = "#99b3ff", 
                               "Female-autosomes" = "darkred", "Female-sex chr." = "#f3a5ae"),
                    labels = c("Female-autosomes" = "Female autosomal", "Female-sex chr." = "Female sex chr.Â´",
                               "Male-autosomes" = "Male autosomal", "Male-sex chr." = "Male sex chr.")){
  if (legend){
    p <- ggplot(data, aes(x = variable, y = nbr_genes, fill = coloring, color = chr))
  } else{
    p <- ggplot(data, aes(x = variable, y = nbr_genes, fill = coloring))
  }
  p <- p + geom_bar(stat = "identity", size = 0)
  p <- p + scale_fill_manual(values = colors, labels = labels)
  p <- p + theme_bw() + labs(x = "", y = "")
  p <- p + theme(legend.position="bottom", legend.title = element_blank(),
                 legend.spacing.y = unit(1, 'mm'),
                 plot.margin = margin(3,3,-5,-5, "mm"),
                 legend.margin = margin(0,0,0,0, "mm"),
                 legend.box.margin = margin(0,0,-10,0, "mm"),
                 axis.title.x = element_text(size = 14),
                 axis.title.y = element_text(size = 14),
                 axis.text.x = element_text(size = 12),
                 axis.text.y = element_text(size = 10),
                 legend.text=element_text(size=12),
                 legend.box.spacing = unit(0,"null"))
   p <- p + guides(fill = "none")
  if (legend){
    p <- p + scale_color_manual(values = c("autosomes" = "black", "sex chr." = "grey"), labels = c("autosomes", "sex chr."), )
    p <- p + guides(fill = "none",
                  color = guide_legend(nrow=nrowLegend, byrow=TRUE, 
                                       override.aes = list(fill = c("black", "grey"))))
  }
  return(p)
}
```

```{r barplot_sexDE, echo=FALSE}
# prepare data
sexChr <- c("chrY", "chrX")
barplot_data_sexDE <- data.frame(variable = c("Female", "Female", "Male", "Male"), 
                           chr = c("autosomes", "sex chr.", "autosomes", "sex chr."), 
                           nbr_genes = c(length(de_results_autosomes$gene_name[which(de_results_autosomes$qvalue < threshold_qvalue & de_results_autosomes$logFC > 0)]),
                                         length(de_results_filtered$gene_name[which((de_results_filtered$chr %in% sexChr) & de_results_filtered$qvalue < threshold_qvalue & de_results_filtered$logFC > 0)]),
                                         length(de_results_autosomes$gene_name[which(de_results_autosomes$qvalue < threshold_qvalue & de_results_autosomes$logFC < 0)]),
                                         length(de_results_filtered$gene_name[which((de_results_filtered$chr %in% sexChr) & de_results_filtered$qvalue < threshold_qvalue & de_results_filtered$logFC < 0)]))
)

barplot_data_sexDE$coloring <- paste(barplot_data_sexDE$variable, barplot_data_sexDE$chr, sep = "-")
barplot_data_sexDE <- barplot_data_sexDE[order(barplot_data_sexDE$nbr_genes, decreasing = TRUE),] # order data
barplot_data_sexDE$coloring <- factor(barplot_data_sexDE$coloring, levels = c("Male-sex chr.", "Female-sex chr.", "Male-autosomes" , "Female-autosomes"))
barplot_data_sexDE$variable <- factor(barplot_data_sexDE$variable, levels = c("Male", "Female"))

# barplot
p_barplot <- barplot(barplot_data_sexDE)
p_barplot
```

## Volcanoplot

```{r volcanoplot_sexDE_auto, echo=FALSE}
keyvals <- rep('grey75', nrow(de_results_autosomes))
names(keyvals) <- rep('NS', nrow(de_results_autosomes))

keyvals[which(de_results_autosomes$logFC < 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'darkblue'
names(keyvals)[which(de_results_autosomes$logFC < 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'Signif. males'

keyvals[which(de_results_autosomes$logFC > 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'darkred'
names(keyvals)[which(de_results_autosomes$logFC > 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'Signif. females'

unique(keyvals)
unique(names(keyvals))

p_volcanoplot <- EnhancedVolcano(de_results_autosomes,
                                 lab = as.character(de_results_autosomes$gene_name),
                                 x = 'logFC',
                                 y = 'qvalue',
                                 xlim = c(-1.5, 1.5),
                                 ylim = c(0,16),
                                 FCcutoff = 2,
                                 pCutoff = 10E-15,
                                 cutoffLineType = 'blank',
                                 colCustom = keyvals,
                                 legendPosition = "none",
                                 ylab = expression('-Log'[10]*' q-value'),
                                 title = NULL,
                                 subtitle = NULL,
                                 caption = NULL)
p_volcanoplot <- p_volcanoplot + theme_bw()
p_volcanoplot <- p_volcanoplot + theme(plot.margin = margin(3,3,3,3, "mm"),
                                       axis.title.x = element_text(size = 14),
                                       axis.title.y = element_text(size = 14),
                                       axis.text.x = element_text(size = 10),
                                       axis.text.y = element_text(size = 10),
                                       panel.border = element_rect(fill=NA, colour = "black", linewidth = 0.2))
p_volcanoplot <- p_volcanoplot + guides(fill = "none", color = "none")

p_volcanoplot
```

## Arrange 2 plots together

```{r figure1C_arrangePlots, echo=FALSE}
ggsave("1C_volcanoplot_barplot_sexDE.svg", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 12, height = 8, units = "cm")
ggsave("1C_volcanoplot_barplot_sexDE.png", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 12, height = 8, units = "cm")
```


# Figure 1 D

Volcanoplot of the pseudotime-DE genes colored by sex (without autosomes). 
And barplot of the number of pseudotime-DE genes

## Load and prepare data

```{r figure1D_load_data, echo=FALSE}
# load pseudotime data
load("data/2-pseudotime/forebrain_pseudotime_analysis.RData")
de_results_pseudotime <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction_pseudotime.txt",
                                    header = TRUE)
```

From the full list of sex-DE genes, remove XIST and the protein-coding genes on the Y-chromosome.

```{r figure1D_remove_training_genes, echo=FALSE}
genes2remove <- c("XIST", as.character(de_results_pseudotime$gene_name[which(de_results_pseudotime$chr == "chrY" & de_results_pseudotime$gene_type == "protein_coding")]))
de_results_pseudotime_filtered <- de_results_pseudotime[which(!de_results_pseudotime$gene_name %in% genes2remove),]
```

From the pseudotime-DE genes remove the genes used to compute the pseudotime.

```{r remove_pseudotime_genes, echo=FALSE}
genes2remove <- rownames(selectedCounts)
de_results_pseudotime_filtered <- de_results_pseudotime_filtered[which(!de_results_pseudotime_filtered$gene_name %in% genes2remove),]
```

## Barplot

```{r barplot_pseudotimeDE, echo=FALSE}
barplot_data_pseudoDE <- data.frame(variable = c("Upregulated", "Downregulated"),
                           nbr_genes = c(length(de_results_pseudotime_filtered$gene_name[which(de_results_pseudotime_filtered$qvalue < threshold_qvalue & de_results_pseudotime_filtered$logFC > 0)]),
                                         length(de_results_pseudotime_filtered$gene_name[which(de_results_pseudotime_filtered$qvalue < threshold_qvalue & de_results_pseudotime_filtered$logFC < 0)]))
)

barplot_data_pseudoDE$coloring <- barplot_data_pseudoDE$variable
barplot_data_pseudoDE <- barplot_data_pseudoDE[order(barplot_data_pseudoDE$nbr_genes, decreasing = TRUE),] # order data

# barplot
p_barplot <- barplot(barplot_data_pseudoDE,
                     colors = c("Upregulated" = 'aquamarine4', "Downregulated" = "chocolate"),
                     labels = c("Upregulated" = 'Up-\nregulated', "Downregulated" = "Down-\nregulated"), 
                     legend = FALSE)
p_barplot <- p_barplot + scale_x_discrete(labels = c("Upregulated" = 'Up-\nregulated', "Downregulated" = "Down-\nregulated"))
p_barplot <- p_barplot + guides(fill = "none")
p_barplot <- p_barplot + theme(plot.margin = margin(3,3,-5,-5, "mm"),
                               axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, size = 12))
p_barplot
```

## Volcanoplot

```{r volcanoplot_pseudotimeDE, echo=FALSE}
keyvals <- rep('grey75', nrow(de_results_pseudotime_filtered))
names(keyvals) <- rep('NS', nrow(de_results_pseudotime_filtered))

keyvals[which(de_results_pseudotime_filtered$logFC < 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'chocolate'
names(keyvals)[which(de_results_pseudotime_filtered$logFC < 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'Signif. Down'

keyvals[which(de_results_pseudotime_filtered$logFC > 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'aquamarine4'
names(keyvals)[which(de_results_pseudotime_filtered$logFC > 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'Signif. Up'

unique(keyvals)
unique(names(keyvals))

p_volcanoplot <- EnhancedVolcano(de_results_pseudotime_filtered,
                                 lab = as.character(de_results_pseudotime_filtered$gene_name),
                                 x = 'logFC',
                                 y = 'qvalue',
                                 xlim = c(-3.5, 3.5),
                                 ylim = c(0,100),
                                 FCcutoff = 10,
                                 pCutoff = 10E-10,
                                 cutoffLineType = 'blank',
                                 colCustom = keyvals,
                                 legendPosition = "none",
                                 ylab = expression('-Log'[10]*' q-value'),
                                 title = NULL,
                                 subtitle = NULL,
                                 caption = NULL)
p_volcanoplot <- p_volcanoplot + theme_bw()
p_volcanoplot <- p_volcanoplot + theme(plot.margin = margin(3,3,3,3, "mm"),
                                       axis.title.x = element_text(size = 14),
                                       axis.title.y = element_text(size = 14),
                                       axis.text.x = element_text(size = 10),
                                       axis.text.y = element_text(size = 10),
                                       panel.border = element_rect(fill=NA, colour = "black", linewidth = 0.2))
p_volcanoplot <- p_volcanoplot + guides(fill = "none", color = "none")

p_volcanoplot
```

## Arrange 2 plots together

```{r figure1D_arrangePlots, echo=FALSE}
ggsave("1D_volcanoplot_barplot_pseudotimeDE.svg", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 12, height = 8, units = "cm")
ggsave("1D_volcanoplot_barplot_pseudotimeDE.png", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 12, height = 8, units = "cm")
```


# Figure 1 E

Upset plot to compare sex-DE and pseudotime-DE genes.

## Prepare data

Data was previously loaded to do figure 1C and 1D.

Merge sexDE and pseudotimeDE data to use it with ComplexUpSet package.

```{r merge_data, echo=FALSE}
merged_data <- merge(de_results_filtered[, c(1:5,11)], de_results_pseudotime_filtered[, c(1:5,11)], #all.x = TRUE,
                     by = c("gene_id", "gene_name", "gene_type", "chr"), suffixes = c("_sexDE", "_pseudotimeDE"))
merged_data$sexDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue, TRUE, FALSE)
merged_data$femaleDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue & merged_data$logFC_sexDE>0, TRUE, FALSE)
merged_data$maleDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue & merged_data$logFC_sexDE<0, TRUE, FALSE)
merged_data$pseudotimeDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue, TRUE, FALSE)
merged_data$upDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue & merged_data$logFC_pseudotimeDE>0, TRUE, FALSE)
merged_data$downDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue & merged_data$logFC_pseudotimeDE<0, TRUE, FALSE)
```

Also add information about autosome/sex-chromosome genes.

```{r add_chr, echo=FALSE}
merged_data$type <- ifelse(merged_data$chr %in% c("chrX", "chrY"), "sex chr.", "autosomes")
merged_data$sexChr <- ifelse(merged_data$chr %in% c("chrX", "chrY"), "TRUE", "FALSE")
merged_data$type <- factor(merged_data$type, levels = c("sex chr.", "autosomes"))
```

Rename columns to improve upset plot.

```{r rename_col, echo=FALSE}
names(merged_data)[c(10, 11, 13, 14)] <- c("female", "male", "up", "down")
```

Format the data for the upset plot from the ComplexHeatmap package.

```{r format_data_upset, echo=FALSE}
combination_matrix <- make_comb_mat(merged_data[c(10, 11, 13, 14)])
combination_matrix <- combination_matrix[comb_degree(combination_matrix) > 0]
```

## Upset plot

```{r upset_plot, echo=FALSE}
colorVal <- list(group = c("up" = "#16967D", "down" = "#D35400", "male" = 'darkblue', "female" = "darkred"))
p <- UpSet(combination_matrix, set_order = c("up", "down", "male", "female"), 
           comb_order = order(comb_size(combination_matrix), decreasing = TRUE),
           row_names_side = "left",
           row_names_gp = gpar(fontsize = 0),
           top_annotation = upset_top_annotation(combination_matrix, 
                                                 show_annotation_name = FALSE,
                                                 add_numbers = TRUE,
                                                 numbers_rot = 0,
                                                 numbers_gp = gpar(fontsize = 10),
                                                 height = unit(5, "cm")),
           left_annotation = rowAnnotation(group = c("female", "male", "down", "up"), 
                                           col = colorVal,
                                           show_annotation_name = FALSE,
                                           show_legend = FALSE)
)
p
# save plot
svg("images/figures/1E_complexUpsetPlot_between_sexDE_pseudotimeDE.svg",
    width = 3.15, height = 3.15)
p
dev.off()
png("images/figures/1E_complexUpsetPlot_between_sexDE_pseudotimeDE.png",
     width = 8, height = 8, units = "cm", res = 30)
p
dev.off()
```


# Figure 1 F

Visualization of example genes.

## Load and prepare data

```{r figure1F_load, echo=FALSE}
# load sex-DE data
load("data/3-DE/normCounts_dream_pseudotime_sva_ind_interaction_phenotypes.RData")
```

## Scatter plots for each example genes

Function to do the scatter plot.

```{r function_plot_gene_pseudotime, echo=FALSE}
plot_Expr_Pseudotime_Sex <- function(expression, metadata, gene, legend = FALSE){
  # extract gene expression
  gene_expr <- expression$E[which(expression$genes$gene_name == gene), ]
  data_plot <- data.frame(expr = gene_expr,
                          pseudotime = metadata$pseudotime,
                          sex = as.factor(metadata$Sex))
  # plot gene expression
  p <- ggplot(data_plot, aes(x = pseudotime, y = expr, color = sex, group = sex))
  p <- p + geom_point()
  p <- p + geom_smooth(method = 'lm', formula = y ~ x)
  p <- p + scale_color_manual(name = NULL, values = c("1" = "darkblue", "2" = "darkred"), labels = c("Male", "Female"))
  p <- p + theme_minimal()
  if (legend){
    p <- p + theme(text = element_text(size=14), legend.position = "inside", legend.position.inside = c(0.9, 0.85))
  } else{
    p <- p + theme(text = element_text(size=14), legend.position="none")
  }
  p <- p + xlab("Pseudotime") + ylab("Gene expression") + ggtitle(NULL)
  return(p)
}
```

Plot the example gene ZFX.

```{r plot_example_gene_ZFX, echo=FALSE}
myGene <- "ZFX"
p <- plot_Expr_Pseudotime_Sex(y, metadata, myGene, legend = TRUE)
p
# save
ggsave(plot = p, filename = "1F_example_gene_ZFX.png", 
       path = "images/figures/", width = 12, height = 8, units = "cm")
ggsave(plot = p, filename = "1F_example_gene_ZFX.svg", 
       path = "images/figures/", width = 12, height = 8, units = "cm")
```


