---
title: "Scripts to create the main figures: Figure 1"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r load_library, echo=FALSE}
suppressPackageStartupMessages( library(tidyverse) )
suppressPackageStartupMessages( library(edgeR) )
suppressPackageStartupMessages( library(limma) )
suppressPackageStartupMessages( library(SummarizedExperiment) )
suppressPackageStartupMessages( library(qvalue) )
suppressPackageStartupMessages( library(EnhancedVolcano) )
suppressPackageStartupMessages( library(gridExtra) )
suppressPackageStartupMessages( library(ComplexHeatmap) )
suppressPackageStartupMessages( library(ggpubr) )
```

```{r setup, echo=FALSE}
knitr::opts_knit$set(root.dir = '~/Desktop/Workspace/HDBR/github-repo/sexDE_prenatal_brain/')
```

```{r param, echo=FALSE}
threshold_qvalue <- 0.01
```


# Figure 1 B

Boxplot of the pseudotime grouped by developmental stages and colored by sex.

## Load data

```{r figure1B_load_data, echo=FALSE}
# load pseudotime data
load("data/2-pseudotime/forebrain_pseudotime_analysis.RData")
```

## Prepare the data

Convert the developmental stages to a continuous variable.

```{r convert2continuous, echo=FALSE}
metadata$DevStageCont <- as.integer(metadata$DevStage)
```

Compute Kendall correlation between the pseudotime and the new continuous developmental stage.

```{r correlation_kendall, echo=FALSE}
corTestRes_Kendall <- cor.test(metadata$DevStageCont, metadata$pseudotime, method = "kendall")
corTestRes_Kendall
```

## Plot 

Scatter plot of the pseudotime and the continuous developmental stage.

```{r scatter_plot, echo=FALSE}
p <- ggplot(metadata, aes(DevStage, pseudotime))
p <- p + geom_boxplot(fill = "#DCDCDC", outlier.shape = NA)
p <- p + theme_minimal() + xlab("Developmental stages")
p <- p + geom_jitter(aes(color = Sex), position=position_jitter(0.2), show.legend = FALSE)
p <- p + annotate("text", x = 1.6, y = 1.1, label = sprintf("tau = %.2f\n p-value = %.2e", corTestRes_Kendall$estimate, corTestRes_Kendall$p.value))
p <- p + scale_color_manual(values = c("1" = "darkblue", "2" = "darkred"))
p <- p + theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
p
# save plot
ggsave("1B_boxplot_devStages_vs_pseudotime.png", 
       path = "images/figures/",
       plot = p, width = 10, height = 8, units = "cm")
```


# Figure 1 C

Volcanoplot of the sex-DE genes colored by sex (without autosomes). 
And barplot of the number of sex-DE genes

## Load and prepare data

```{r figure1C_load_data, echo=FALSE}
# load sex-DE data
load("data/3-DE/normCounts_dream_pseudotime_sva_ind_interaction_phenotypes.RData")
de_results <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction.txt",
                         header = TRUE)
```

From the full list of sex-DE genes, remove XIST and the protein-coding genes on the Y-chromosome.

```{r remove_training_genes, echo=FALSE}
genes2remove <- c("XIST", as.character(de_results$gene_name[which(de_results$chr == "chrY" & de_results$gene_type == "protein_coding")]))
de_results_filtered <- de_results[which(!de_results$gene_name %in% genes2remove),]
```

Create a data frame containing only the results for the autosomes.

```{r autosomes, echo=FALSE}
chrNonAuto <- c("chrY", "chrX", "chrM")
de_results_autosomes <- de_results_filtered[which(!de_results_filtered$chr %in% chrNonAuto),]
```

## Barplot

```{r function_barplot, echo=FALSE}
barplot <- function(data, legend = TRUE, nrowLegend = 4,
                    colors = c("Male-other" = "#053061", "Male-sex" = "#99b3ff", 
                               "Female-other" = "#B2182B", "Female-sex" = "#f3a5ae"),
                    labels = c("Female-other" = "Female autosomal", "Female-sex" = "Female sex chr.Â´",
                               "Male-other" = "Male autosomal", "Male-sex" = "Male sex chr.")){
  p <- ggplot(data, aes(x = variable, y = nbr_genes, fill = coloring))
  p <- p + geom_bar(stat = "identity")
  p <- p + scale_fill_manual(values = colors, labels = labels)
  p <- p + theme_bw() + labs(x = "", y = "")
  p <- p + theme(legend.position="bottom", legend.title = element_blank()) 
  p <- p + guides(fill = guide_legend(nrow=nrowLegend, byrow=TRUE))
  if (!legend){
    p <- p +  guides(fill = "none")
  }
  return(p)
}
```

```{r barplot_sexDE, echo=FALSE}
# prepare data
sexChr <- c("chrY", "chrX")
barplot_data_sexDE <- data.frame(variable = c("Female", "Female", "Male", "Male"), 
                           chr = c("other", "sex", "other", "sex"), 
                           nbr_genes = c(length(de_results_autosomes$gene_name[which(de_results_autosomes$qvalue < threshold_qvalue & de_results_autosomes$logFC > 0)]),
                                         length(de_results_filtered$gene_name[which((de_results_filtered$chr %in% sexChr) & de_results_filtered$qvalue < threshold_qvalue & de_results_filtered$logFC > 0)]),
                                         length(de_results_autosomes$gene_name[which(de_results_autosomes$qvalue < threshold_qvalue & de_results_autosomes$logFC < 0)]),
                                         length(de_results_filtered$gene_name[which((de_results_filtered$chr %in% sexChr) & de_results_filtered$qvalue < threshold_qvalue & de_results_filtered$logFC < 0)]))
)

barplot_data_sexDE$coloring <- paste(barplot_data_sexDE$variable, barplot_data_sexDE$chr, sep = "-")
barplot_data_sexDE <- barplot_data_sexDE[order(barplot_data_sexDE$nbr_genes, decreasing = TRUE),] # order data
barplot_data_sexDE$coloring <- factor(barplot_data_sexDE$coloring, levels = c("Male-sex", "Female-sex", "Male-other" , "Female-other"))
barplot_data_sexDE$variable <- factor(barplot_data_sexDE$variable, levels = c("Male", "Female"))

# barplot
p_barplot <- barplot(barplot_data_sexDE)
p_barplot
# save plot
# ggsave(plot = "1C_barplot_nbr_sexDE_genes.png", 
#        path = 'images/figures/',
#        namePlot, width = 8, height = 17.5, units = "cm")
```

## Volcanoplot

```{r volcanoplot_sexDE_auto, echo=FALSE}
keyvals <- rep('grey75', nrow(de_results_autosomes))
names(keyvals) <- rep('NS', nrow(de_results_autosomes))

keyvals[which(de_results_autosomes$logFC < 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'darkblue'
names(keyvals)[which(de_results_autosomes$logFC < 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'Signif. males'

keyvals[which(de_results_autosomes$logFC > 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'darkred'
names(keyvals)[which(de_results_autosomes$logFC > 0 & de_results_autosomes$qvalue < threshold_qvalue)] <- 'Signif. females'

unique(keyvals)
unique(names(keyvals))

p_volcanoplot <- EnhancedVolcano(de_results_autosomes,
                                 lab = as.character(de_results_autosomes$gene_name),
                                 x = 'logFC',
                                 y = 'qvalue',
                                 xlim = c(-2.5, 2.5),
                                 ylim = c(0,20),
                                 FCcutoff = 2,
                                 pCutoff = 10E-15,
                                 cutoffLineType = 'blank',
                                 colCustom = keyvals,
                                 legendPosition = "none",
                                 ylab = expression('-Log'[10]*' q-value'),
                                 title = NULL,
                                 subtitle = NULL,
                                 caption = NULL)
p_volcanoplot

# save plot
# ggsave('1C_volcanoplot_colored_HDBR_dream_forebrain_pseudotime_sva_ind_interaction_sexDE_auto.png', 
#        path = "images/figures/",
#        width = 20, height = 15, units = "cm")
```

## Arrange 2 plots together

```{r figure1C_arrangePlots, echo=FALSE}
ggsave("1C_volcanoplot_barplot_sexDE.png", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 16, height = 10, units = "cm")
```


# Figure 1 D

Volcanoplot of the pseudotime-DE genes colored by sex (without autosomes). 
And barplot of the number of pseudotime-DE genes

## Load and prepare data

```{r figure1D_load_data, echo=FALSE}
# load pseudotime data
load("data/2-pseudotime/forebrain_pseudotime_analysis.RData")
de_results_pseudotime <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction_pseudotime.txt",
                                    header = TRUE)
```

From the full list of sex-DE genes, remove XIST and the protein-coding genes on the Y-chromosome.

```{r figure1D_remove_training_genes, echo=FALSE}
genes2remove <- c("XIST", as.character(de_results_pseudotime$gene_name[which(de_results_pseudotime$chr == "chrY" & de_results_pseudotime$gene_type == "protein_coding")]))
de_results_pseudotime_filtered <- de_results_pseudotime[which(!de_results_pseudotime$gene_name %in% genes2remove),]
```

From the pseudotime-DE genes remove the genes used to compute the pseudotime.

```{r remove_pseudotime_genes, echo=FALSE}
genes2remove <- rownames(selectedCounts)
de_results_pseudotime_filtered <- de_results_pseudotime_filtered[which(!de_results_pseudotime_filtered$gene_name %in% genes2remove),]
```

## Barplot

```{r barplot_pseudotimeDE, echo=FALSE}
barplot_data_pseudoDE <- data.frame(variable = c("Up", "Down"),
                           nbr_genes = c(length(de_results_pseudotime_filtered$gene_name[which(de_results_pseudotime_filtered$qvalue < threshold_qvalue & de_results_pseudotime_filtered$logFC > 0)]),
                                         length(de_results_pseudotime_filtered$gene_name[which(de_results_pseudotime_filtered$qvalue < threshold_qvalue & de_results_pseudotime_filtered$logFC < 0)]))
)

barplot_data_pseudoDE$coloring <- barplot_data_pseudoDE$variable
barplot_data_pseudoDE <- barplot_data_pseudoDE[order(barplot_data_pseudoDE$nbr_genes, decreasing = TRUE),] # order data

# barplot
p_barplot <- barplot(barplot_data_pseudoDE,
                     colors = c("Up" = 'aquamarine4', "Down" = "chocolate"),
                     labels = c("Up" = 'Upregulated', "Down" = "Downregulated"), 
                     nrowLegend = 2)
p_barplot
# save plot
# ggsave(plot = "1D_barplot_nbr_pseudotimeDE_genes.png", 
#        path = "images/figures/", 
#        namePlot, width = 8, height = 16, units = "cm")
```

## Volcanoplot

```{r volcanoplot_pseudotimeDE, echo=FALSE}
keyvals <- rep('grey75', nrow(de_results_pseudotime_filtered))
names(keyvals) <- rep('NS', nrow(de_results_pseudotime_filtered))

keyvals[which(de_results_pseudotime_filtered$logFC < 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'chocolate'
names(keyvals)[which(de_results_pseudotime_filtered$logFC < 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'Signif. Down'

keyvals[which(de_results_pseudotime_filtered$logFC > 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'aquamarine4'
names(keyvals)[which(de_results_pseudotime_filtered$logFC > 0 & de_results_pseudotime_filtered$qvalue < threshold_qvalue)] <- 'Signif. Up'

unique(keyvals)
unique(names(keyvals))

p_volcanoplot <- EnhancedVolcano(de_results_pseudotime_filtered,
                                 lab = as.character(de_results_pseudotime_filtered$gene_name),
                                 x = 'logFC',
                                 y = 'qvalue',
                                 FCcutoff = 10,
                                 pCutoff = 10E-10,
                                 cutoffLineType = 'blank',
                                 colCustom = keyvals,
                                 legendPosition = "none",
                                 ylab = expression('-Log'[10]*' q-value'),
                                 title = NULL,
                                 subtitle = NULL,
                                 caption = NULL)
p_volcanoplot

# save plot
# ggsave('1D_volcanoplot_colored_HDBR_dream_forebrain_pseudotime_sva_ind_interaction_pseudotimeDE.png', 
#        path = "images/figures/",
#        width = 12, height = 15, units = "cm")
```

## Arrange 2 plots together

```{r figure1D_arrangePlots, echo=FALSE}
ggsave("1D_volcanoplot_barplot_pseudotimeDE.png", 
       path = "images/figures/",
       arrangeGrob(p_volcanoplot, p_barplot, ncol = 2, widths=c(4,2)),
       width = 16, height = 10, units = "cm")
```


# Figure 1 E

Upset plot to compare sex-DE and pseudotime-DE genes.

## Prepare data

Data was previously loaded to do figure 1C and 1D.

Merge sexDE and pseudotimeDE data to use it with ComplexUpSet package.

```{r merge_data, echo=FALSE}
merged_data <- merge(de_results_filtered[, c(1:5,11)], de_results_pseudotime_filtered[, c(1:5,11)], #all.x = TRUE,
                     by = c("gene_id", "gene_name", "gene_type", "chr"), suffixes = c("_sexDE", "_pseudotimeDE"))
merged_data$sexDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue, TRUE, FALSE)
merged_data$femaleDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue & merged_data$logFC_sexDE>0, TRUE, FALSE)
merged_data$maleDE <- ifelse(merged_data$qvalue_sexDE<threshold_qvalue & merged_data$logFC_sexDE<0, TRUE, FALSE)
merged_data$pseudotimeDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue, TRUE, FALSE)
merged_data$upDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue & merged_data$logFC_pseudotimeDE>0, TRUE, FALSE)
merged_data$downDE <- ifelse(merged_data$qvalue_pseudotimeDE<threshold_qvalue & merged_data$logFC_pseudotimeDE<0, TRUE, FALSE)
```

Also add information about autosome/sex-chromosome genes.

```{r add_chr, echo=FALSE}
merged_data$type <- ifelse(merged_data$chr %in% c("chrX", "chrY"), "sex chr.", "autosomes")
merged_data$sexChr <- ifelse(merged_data$chr %in% c("chrX", "chrY"), "TRUE", "FALSE")
merged_data$type <- factor(merged_data$type, levels = c("sex chr.", "autosomes"))
```

Rename columns to improve upset plot.

```{r rename_col, echo=FALSE}
names(merged_data)[c(10, 11, 13, 14)] <- c("female", "male", "up", "down")
```

Format the data for the upset plot from the ComplexHeatmap package.

```{r format_data_upset, echo=FALSE}
combination_matrix <- make_comb_mat(merged_data[c(10, 11, 13, 14)])
```

## Upset plot

```{r upset_plot, echo=FALSE}
colorVal <- list(group = c("up" = "#16967D", "down" = "#D35400", "male" = "#053061", "female" = "#B2182B"))
p <- UpSet(combination_matrix, set_order = c("up", "down", "male", "female"), 
           comb_order = order(comb_size(combination_matrix), decreasing = TRUE),
           row_names_side = "left",
           top_annotation = upset_top_annotation(combination_matrix, 
                                                 show_annotation_name = FALSE,
                                                 add_numbers = TRUE,
                                                 numbers_rot = 0,
                                                 numbers_gp = gpar(fontsize = 8),
                                                 height = unit(5, "cm")),
           left_annotation = rowAnnotation(group = c("female", "male", "down", "up"), 
                                            col = colorVal,
                                            show_annotation_name = FALSE,
                                            show_legend = FALSE)
)
p
# save plot
png("images/figures/1E_complexUpsetPlot_between_sexDE_pseudotimeDE.png",
    width = 8, height = 8, units = "cm", res = 300)
p
dev.off()
```


# Figure 1 F

Visualization of example genes.

## Load and prepare data

```{r figure1F_load, echo=FALSE}
# load sex-DE data
load("data/3-DE/normCounts_dream_pseudotime_sva_ind_interaction_phenotypes.RData")
de_results <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction.txt",
                         header = TRUE)
de_results_interaction <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction_term.txt", 
                                     header = TRUE)
de_results_pseudotime <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction_pseudotime.txt",
                                    header = TRUE)
```

## Scatter plots for each example genes

Function to do the scatter plot.

```{r function_plot_gene_pseudotime, echo=FALSE}
plot_Expr_Pseudotime_Sex <- function(expression, metadata, gene, DEinfo, xtext = 0.5, ytext = 6, legend = FALSE){
  # extract gene expression
  gene_expr <- expression$E[which(expression$genes$gene_name == gene), ]
  data_plot <- data.frame(expr = gene_expr,
                          pseudotime = metadata$pseudotime,
                          sex = as.factor(metadata$Sex))
  # text to add to the plot
  textLogFC <- paste(paste0("logFC sex-DE = ", round(DEinfo$sex, digits = 2)),
                     paste0("logFC pseudotime-DE = ", round(DEinfo$pseudotime, digits = 2)),
                     paste0("logFC interaction-DE = ", round(DEinfo$interaction, digits = 2)), 
                     sep = "\n")
  # plot gene expression
  p <- ggplot(data_plot, aes(x = pseudotime, y = expr, color = sex, group = sex))
  p <- p + geom_point()
  p <- p + geom_smooth(method = 'lm', formula = y ~ x)
  p <- p + scale_color_manual(name = "Sex", values = c("1" = "#1565c1", "2" = "#B2182B"), labels = c("Male", "Female"))
  p <- p + theme_minimal()
  if (legend){
    p <- p + theme(text = element_text(size=20))
  } else{
    p <- p + theme(text = element_text(size=20), legend.position="none")
  }
  p <- p + annotate(geom = "text", x = xtext, y = ytext, label = textLogFC, hjust = "left", vjust = "top", size = 5)
  p <- p + xlab("Pseudotime") + ylab("Gene expression") + ggtitle(gene)
  return(p)
}
```

Plots the 5 example genes.

```{r plot_exemple_genes, echo=FALSE}
example_genes <- c("SERP2", "TTPA", "DOCK1", "BLK", "DCX")
x_text_list <- c(-0.25, -1.01, -0.9, -1.01, -0.3)
y_text_list <- c(2.7, -2, 3, -4, 7.5)
legendBool <- c(TRUE, FALSE, FALSE, FALSE, FALSE)
plots <- list()
for (i in 1:5){
  myGene <- example_genes[i]
  x_text <- x_text_list[i]
  y_text <- y_text_list[i]
  DEinfo <- list("sex" = de_results$logFC[which(de_results$gene_name == myGene)], 
                 "pseudotime" = de_results_pseudotime$logFC[which(de_results_pseudotime$gene_name == myGene)], 
                 "interaction" = de_results_interaction$logFC[which(de_results_interaction$gene_name == myGene)])
  p <- plot_Expr_Pseudotime_Sex(y, metadata, myGene, DEinfo = DEinfo, xtext = x_text, ytext = y_text, legend = legendBool[i])
  plots[[i]] <- p
}
```

Arrange the plots together and save the results.

```{r figure1F_arrangePlots, echo=FALSE}
legend <- as_ggplot(get_legend(plots[[1]])) # extract legend from first plot
plots[[1]] <- plots[[1]] + theme(legend.position="none") # remove legend from this same plot
layout_matrix <- matrix(c(1, 1, 2, 2, 3, 3, 7, 4, 4, 5, 5, 6), nrow = 2, byrow = TRUE)

ggsave("1F_example_genes.png", 
       path = "images/figures/",
       grid.arrange(plots[[1]], plots[[2]], plots[[3]], plots[[4]], plots[[5]], legend, layout_matrix = layout_matrix),
       width = 36, height = 24, units = "cm")
```


