---
title: "Pseudotime analysis with the developmental stages as covariates"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r library, echo=FALSE}
suppressPackageStartupMessages(library(GenomicFeatures))
suppressPackageStartupMessages(library(edgeR))
suppressPackageStartupMessages(library(limma))
suppressPackageStartupMessages(library(SummarizedExperiment))
suppressPackageStartupMessages(library(phenopath))
suppressPackageStartupMessages(library(ade4))
suppressPackageStartupMessages(library(factoextra))
suppressPackageStartupMessages(library(ggplot2))
```

```{r setup, echo=FALSE}
knitr::opts_knit$set(root.dir = '~/Desktop/Workspace/HDBR/github-repo/sexDE_prenatal_brain/')
```

# Load and prepare data

## Load the counts and the phenotype data

```{r load_data, echo=FALSE}
load("data/1-preprocessing/counts_forebrain_phenotypes.RData")
```


## Prepare metadata

```{r prepare_metadata, echo=FALSE}
Sex <- as.factor(phenotypes$assigned_sex)
DevStage <- phenotypes$newDevStage
Individual <- as.factor(phenotypes$ind)
ForebrainRegion <- phenotypes$newBrainRegion
Batch <- phenotypes$batch
# metadata
metadata <- data.frame(Sex = Sex, DevStage = DevStage, Individual = Individual, ForebrainRegion = ForebrainRegion, Batch = Batch)
rownames(metadata) <- phenotypes$experiment_name
```



## Convert counts in TPM values

Load the annotation file.

```{r load_annotation, echo=FALSE}
# First, import the GTF-file that you have also used as input for htseq-count
txdb <- txdbmaker::makeTxDbFromGFF("data/0-input/gencode.v28.annotation.gff3",format="gff")

# then collect the exons per gene id
exons.list.per.gene <- exonsBy(txdb,by="gene")

# then for each gene, reduce all the exons to a set of non overlapping exons, calculate their lengths (widths) and sum them
exonic.gene.sizes <-  sum(width(reduce(exons.list.per.gene)))
#names(exonic.gene.sizes)=matrix(unlist(strsplit(as.character(names(exonic.gene.sizes)),"\\.")),ncol=2,byrow=T)[,1]
names(exonic.gene.sizes) <- names(exonic.gene.sizes)

gtf.gene.length_df <- cbind(data.frame(names(exonic.gene.sizes)),as.numeric(exonic.gene.sizes))
names(gtf.gene.length_df) <- c("gene_id","size")
gtf.gene.length_df <- unique(gtf.gene.length_df)

# Merge gene size with gene metadata
gene_metadata <- merge(x$genes, gtf.gene.length_df)
x$genes <- gene_metadata
```

Transform the counts data in log(TPM+1).

```{r transform_counts, echo=FALSE}
counts2tpm <- function(counts, genes){
  # RPK = count/geneLength(kb)
  counts <- counts/(genes$size*1e-3)
  # Sum of RPK for each sample
  library_depth <- colSums(counts)
  # TPM = RPK/(sum(RPK)/1e6)
  counts <- sweep(counts, 2, library_depth/1e6,`/`)
  return(counts)
}
tpm <- counts2tpm(x$counts, x$genes)
logTpm <- log2(tpm+1)
rownames(logTpm) <- x$genes$gene_name
```



## Format data in a SummarizedExperiment object

Select the genes to use in the phenopath analysis.

```{r select_genes, echo=FALSE}
# select higly variable genes
variance <- rowVars(logTpm)
ntop <- min(2000, dim(logTpm)[1])
selectntop <- order(variance, decreasing=TRUE)[seq_len(ntop)]
selectedCounts <- logTpm[selectntop,]
```

Create the SummarizedExperiment object

```{r summarizedExperiment_obj, echo=FALSE}
# initialize sce object
sce <- SummarizedExperiment(assays = list(exprs = selectedCounts), 
                            colData = metadata)
sce
```


# Phenopath analysis

## With Sex and Developmental stages

```{r fit_phenopath_model, echo=FALSE}
x_mat <- model.matrix(~ Sex + DevStage)
fit <- phenopath(sce, x_mat, elbo_tol = 1e-6, thin = 20)
```

Elbo plot

```{r elbo_plot, echo=FALSE}
p <- plot_elbo(fit)
p
# save plot
ggsave("Elbo_plot.png",
       path = "images/2-pseudotime/analysis_withDevStage/",
       plot = p, width = 12, height = 10, units = "cm")
```

Summary of the pseudotime.

```{r summary_pseudotime, echo=FALSE}
summary(trajectory(fit))
```



Add pseudotime to metadata.

```{r infered_pseudotime, echo=FALSE}
metadata$pseudotime <- trajectory(fit)
```

## Save the pseudotime results in a RData

```{r save_pseudotime_res, echo=FALSE}
save(fit, sce, x_mat, selectedCounts, metadata, file = "data/2-pseudotime/analysis_withDevStage/forebrain_pseudotime_analysis_withDevStage.RData")
```

```{r save_metadata, echo=FALSE}
save(metadata, file = "data/2-pseudotime/analysis_withDevStage/forebrain_metadata_pseudotime_withDevStage.RData")
```


# Understand the genes used to compute the pseudotime

The "m_z" from the "fit" abject gives us the pseudotime values.
But which variable from the "fit" object gives us more information about the importance of the genes for the pseudotime?


# Compare this pseudotime with the one from the main analysis

We compare this new pseudotime with the one used in the main analysis (without developmental stages as a covariate).

## Load pseudotime main analysis

Load the original pseudotime values.

```{r load_metadata_mainAnalysis, echo=FALSE}
loadRData <- function(fileName){
  #loads an RData file, and returns it
  load(fileName)
  get(ls()[ls() != "fileName"])
}
metadata_main <- loadRData("data/2-pseudotime/forebrain_metadata_pseudotime.RData")
```


## Check correlation between the 2 pseudotimes

Compute the correlation between this new pseudotime and the original pseudotime variables.

```{r corr_pseudotimes, echo=FALSE}
cor.test(metadata_main$pseudotime, metadata$pseudotime)
```

Do a scatter plot of the 2 pseudotime.

```{r scatter_plot_pseudotime, echo=FALSE}
# merge data
data2plot <- merge(metadata_main[, c(1,2,6)], metadata[, c(1,2,6)], by = c("row.names", "Sex", "DevStage"), suffixes = c("_original", "_new"))
# plot
p <- ggplot(data2plot, aes(x = pseudotime_original, y = pseudotime_new))
p <- p + geom_point()
p <- p + theme_minimal()
p
```

The 2 computed pseudotimes correlate pretty well (r2 = 0.76).


## Check correlation of the new pseudotime with developmental stages

Convert the developmental stages to a continuous variable.

```{r convert2continuous, echo=FALSE}
metadata$DevStageCont <- as.integer(metadata$DevStage)
```

Compute Kendall correlation between the pseudotime and the new continuous developmental stage.

```{r correlation_kendall, echo=FALSE}
corTestRes_Kendall <- cor.test(metadata$DevStageCont, metadata$pseudotime, method = "kendall")
corTestRes_Kendall
```

And also do a boxplot + scatter plot of the pseudotime and the continuous developmental stage.

```{r scatter_plot, echo=FALSE}
p <- ggplot(metadata, aes(DevStage, pseudotime))
p <- p + geom_boxplot(fill = "#DCDCDC", outlier.shape = NA)
p <- p + theme_minimal() + xlab("Developmental stages")
p <- p + geom_jitter(aes(color = Sex), position=position_jitter(0.2), show.legend = FALSE)
p <- p + annotate("text", x = 1.5, y = 1.1, label = sprintf("tau = %.2f\n p-value = %.2e", corTestRes_Kendall$estimate, corTestRes_Kendall$p.value))
p <- p + scale_color_manual(values = c("1" = "darkblue", "2" = "darkred"))
p
# save plot
ggsave("boxplot_devStages_vs_pseudotime.png", 
       path = "images/2-pseudotime/analysis_withDevStage/",
       plot = p, width = 15, height = 10, units = "cm")
```



# Session Info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```