---
title: "MetaBrain network enrichment analysis with hypergeometric test"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r load_library, echo=FALSE}
suppressPackageStartupMessages( library(edgeR) )
suppressPackageStartupMessages( library(limma) )
suppressPackageStartupMessages( library(qvalue) )
suppressPackageStartupMessages( library(variancePartition) )
suppressPackageStartupMessages( library(tibble) )
suppressPackageStartupMessages( library(tidyverse) )
suppressPackageStartupMessages( library(ggnewscale) )
suppressPackageStartupMessages( library(fgsea) )
suppressPackageStartupMessages( library(gridExtra) )
suppressPackageStartupMessages( library(data.table) )
suppressPackageStartupMessages( library(fastmatch) )
suppressPackageStartupMessages( library(scales) )
suppressPackageStartupMessages( library(readxl) )
```

```{r, setup, include=FALSE, echo=FALSE}
knitr::opts_knit$set(root.dir = '~/Desktop/Workspace/HDBR/github-repo/sexDE_prenatal_brain/')
```

# Load data and functions

## Load functions for GSEA analysis with control

```{r load_gsea_withCtrl_function, echo=FALSE}
source("scripts/analyses/5-postprocessing/4-DiseaseEnrichment/enrichment_hyper.R")
```


## Create genesets from metabrain data

Load dowstreamer results from the metaBrain paper for each diseases: ALS, Parkinson, Alzheimer, Schizophrenia and MS.
And create genesets from the loaded data.

```{r load_downstreamer_function, echo=FALSE}
load_downstreamer <- function(file, sheetName){
  my_data <- read_excel(file, sheet = sheetName)
  return(my_data)
}
```

```{r generate_geneset_function, echo=FALSE}
generate_geneset <- function(myData){
  return(myData$`Gene ID`[which(myData$`FDR 5% significant` == "TRUE")])
}
```


```{r loadNformat_downstreamer_data, echo=FALSE}
excel_files <- c("data/5-postprocessing/6-Network/deKlein_2023_MetaBrain/Supplementary Table 25 - Downstreamer - ALS EUR_ASIA enrichments.xlsx",
                 "data/5-postprocessing/6-Network/deKlein_2023_MetaBrain/Supplementary Table 26 - Downstreamer - Parkinsons Nalls2019 enrichments.xlsx",
                 "data/5-postprocessing/6-Network/deKlein_2023_MetaBrain/Supplementary Table 27 - Downstreamer - Schizophrenia Ripke2014 enrichments.xlsx",
                 "data/5-postprocessing/6-Network/deKlein_2023_MetaBrain/Supplementary Table 28 - Downstreamer - Alzheimers Kunkle enrichments.xlsx",
                 "data/5-postprocessing/6-Network/deKlein_2023_MetaBrain/Supplementary Table 29 - Downstreamer - Multiple sclerosis Patsopoulos enrichments.xlsx")
sheet_names <- c("GenePrioritization_MetaBrain", "GenePrioritization_MetaBrainCor", "GenePrioritization_MetaBrainCer")
traits <- c("ALS", "Parkinsons", "Schizophrenia", "Alzheimer", "MS")
datasets <- c("MetaBrain", "MetaBrainCortex", "MetaBrainCerebellum")
# for each trait, load the 3 downstreamer results
downstreamer_data <- list()
downstreamer_genesets <- list()
i <- 1
for (myFile in excel_files){
  j <- 1
  for (mySheet in sheet_names){
    myTrait <- traits[i]
    myData <- datasets[j]
    loaded_data <- load_downstreamer(myFile, mySheet)
    downstreamer_data[[myTrait]][[myData]] <- loaded_data
    myGeneset <- generate_geneset(loaded_data)
    downstreamer_genesets[[myTrait]][[myData]] <- myGeneset
    j <- j + 1
  }
  i <- i + 1
}
```

Filtered the created genesets to keep only the ones with at least 5 genes.

```{r filter_genesets, echo=FALSE}
downstreamer_genesets_unlist <- unlist(downstreamer_genesets, recursive = FALSE)
downstreamer_data_unlist <- unlist(downstreamer_data, recursive = FALSE)
genesets_length <- lapply(downstreamer_genesets_unlist, length)
downstreamer_genesets_filtered <- list()
downstreamer_data_filtered <- list()
for (geneset in names(downstreamer_genesets_unlist)){
  if (genesets_length[[geneset]] >= 10){
    downstreamer_genesets_filtered[[geneset]] <- downstreamer_genesets_unlist[[geneset]]
    downstreamer_data_filtered[[geneset]] <- downstreamer_data_unlist[[geneset]]
  }
}
```

Get the list of genes analyzed in every downstreamer analysis used in the enrichment analysis.

```{r background_downstreamer, echo=FALSE}
analyses <- strsplit(names(downstreamer_genesets_filtered), split = "[.]")
# for each analysis, get all ensembl gene id
background_tmp <- c()
for (i in 1:length(analyses)){
  background_tmp <- c(background_tmp, downstreamer_data_filtered[[i]]$`Gene ID`)
}
background <- unique(background_tmp)
```



## Load constraint data

LOEUF score

```{r load_LOEUF, echo=FALSE}
# load loeuf score
constraint_data <- read.table("data/5-postprocessing/1-GeneProperties/gnomad.v4.1.constraint_metrics.tsv", 
                              header = TRUE, sep = "\t")
constraint_data <- constraint_data[c("gene_id", "canonical", "mane_select", "lof.oe_ci.upper")]
# keep MANE transcript or canonical if no MANE transcript for a gene
constraint_data <- constraint_data[which((constraint_data$mane_select == "true" | constraint_data$canonical == "true") & startsWith(constraint_data$gene_id, "ENSG")), ]
constraint_data <- constraint_data[c("gene_id", "lof.oe_ci.upper")]
names(constraint_data) <- c("gene_id", "LOEUF")
```


## Load DE and bayes data

For each analysis, keep only the genes that were analyzed in the metaBrain analysis.

### HDBR

Load HDBR sex-DE data.

```{r load_data_HDBR_sexDE, echo=FALSE}
hdbr_sexDE_results <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction.txt",
                         header = TRUE, check.names=FALSE)
# remove genes used for sex assignment
genes2remove <- c("XIST", as.character(hdbr_sexDE_results$gene_name[which(hdbr_sexDE_results$chr == "chrY" & hdbr_sexDE_results$gene_type == "protein_coding")]))
hdbr_sexDE_results <- hdbr_sexDE_results[which(!hdbr_sexDE_results$gene_name %in% genes2remove),]
# format
hdbr_sexDE_results$ensemblID <- sapply(strsplit(as.character(hdbr_sexDE_results$gene_id), "\\."), function(x){x[1]})
hdbr_sexDE_results$exprbin <- cut(hdbr_sexDE_results$AveExpr, quantile(hdbr_sexDE_results$AveExpr, seq(0,1,0.1)), 
                                  labels=c(1:10), include.lowest=T) # add expression bins
hdbr_sexDE_results_filtered <- hdbr_sexDE_results[which(hdbr_sexDE_results$ensemblID %in% background), ]
```

Load HDBR pseudotime-DE data.

```{r load_data_HDBR_pseudotimeDE, echo=FALSE}
hdbr_pseudotimeDE_results <- read.table("data/3-DE/VoomDream_topTable_forebrain_pseudotime_sva_ind_interaction_pseudotime.txt",
                         header = TRUE, check.names=FALSE)
# remove genes used to infer pseudotime
load("data/2-pseudotime/forebrain_pseudotime_analysis.RData")
genes2remove <- rownames(selectedCounts)
hdbr_pseudotimeDE_results <- hdbr_pseudotimeDE_results[which(!hdbr_pseudotimeDE_results$gene_name %in% genes2remove),]
# format
hdbr_pseudotimeDE_results$ensemblID <- sapply(strsplit(as.character(hdbr_pseudotimeDE_results$gene_id), "\\."), function(x){x[1]})
hdbr_pseudotimeDE_results$exprbin <- cut(hdbr_pseudotimeDE_results$AveExpr, quantile(hdbr_pseudotimeDE_results$AveExpr, seq(0,1,0.1)), 
                                         labels=c(1:10), include.lowest=T) # add expression bins
hdbr_pseudotimeDE_results_filtered <- hdbr_pseudotimeDE_results[which(hdbr_pseudotimeDE_results$ensemblID %in% background), ]
```


### GTEx

Load GTEx sex-DE data.

```{r load_data_GTEx_sexDE, echo=FALSE}
gtex_sexDE_results <- read.table("data/4-GTEx/1-GTEx_analysis/brain_forebrain_v8_VoomDream_topTable_with_sva_ind_pseudotime_interaction.txt",
                         header = TRUE, check.names=FALSE)
# remove genes used for sex assignment
genes2remove <- c("XIST", as.character(gtex_sexDE_results$gene_name[which(gtex_sexDE_results$chr == "chrY" & gtex_sexDE_results$gene_type == "protein_coding")]))
gtex_sexDE_results <- gtex_sexDE_results[which(!gtex_sexDE_results$gene_name %in% genes2remove),]
# format
gtex_sexDE_results$ensemblID <- sapply(strsplit(as.character(gtex_sexDE_results$gene_id), "\\."), function(x){x[1]})
gtex_sexDE_results$exprbin <- cut(gtex_sexDE_results$AveExpr, quantile(gtex_sexDE_results$AveExpr, seq(0,1,0.1)), 
                                  labels=c(1:10), include.lowest=T) # add expression bins
gtex_sexDE_results_filtered <- gtex_sexDE_results[which(gtex_sexDE_results$ensemblID %in% background), ]
```


Load GTEx pseudotime-DE data.

```{r load_data_GTEx_pseudotimeDE, echo=FALSE}
gtex_pseudotimeDE_results <- read.table("data/4-GTEx/1-GTEx_analysis/brain_forebrain_v8_VoomDream_topTable_with_sva_ind_pseudotime_interaction_pseudotime.txt",
                         header = TRUE, check.names=FALSE)
# remove genes used to infer pseudotime
load("data/4-GTEx/1-GTEx_analysis/GTEx_v8_forebrain_pseudotime_analysis.RData")
genes2remove <- rownames(selectedCounts)
gtex_pseudotimeDE_results <- gtex_pseudotimeDE_results[which(!gtex_pseudotimeDE_results$gene_name %in% genes2remove),]
# format
gtex_pseudotimeDE_results$ensemblID <- sapply(strsplit(as.character(gtex_pseudotimeDE_results$gene_id), "\\."), function(x){x[1]})
gtex_pseudotimeDE_results$exprbin <- cut(gtex_pseudotimeDE_results$AveExpr, quantile(gtex_pseudotimeDE_results$AveExpr, seq(0,1,0.1)), 
                                         labels=c(1:10), include.lowest=T) # add expression bins
gtex_pseudotimeDE_results_filtered <- gtex_pseudotimeDE_results[which(gtex_pseudotimeDE_results$ensemblID %in% background), ]
```


### Bayesian score

```{r load_bayes_results, echo=FALSE}
bayesResults <- read.table("/Users/benoicla/Desktop/Workspace/HDBR/sexDE_prenatal_brain/data/4-postprocessing/4-Comparison_GTEx/sexDE_specificity/lineModel_prenatalVSadult_6models.txt", header = TRUE)
# merge bayes results with prenatal sex-DE results
bayesResults_merged <- merge(bayesResults, hdbr_sexDE_results, by = "ensemblID", suffixes = c("", "_DE"))
# format specificity variable
bayesResults_merged$specificity <- factor(bayesResults_merged$specificity,
                                          levels = c("ADULT", "OPPOSITE", "PRENATAL", "SHARED_UNSPECIFIED", 
                                                     "SHARED0.5", "SHARED1", "SHARED2", "UNCLASSIFIED"),
                                          labels = c("adult_specific", "opposite", "prenatal_specific", "shared", 
                                                     "shared", "shared", "shared", "unclassified"))
bayesResults_merged_filtered <- bayesResults_merged[which(bayesResults_merged$ensemblID %in% background), ]
```



# Enrichment analyses

## Launch enrichment analyses

Number of enrichment analyses to do:
- 6 datasets: HDBR (sex-DE & pseudotime-DE), GTEx (sex-DE & pseudotime-DE), shared and prenatal specific 
- 3 directions: both, female-biased and male-biased
- with or without controlling for gene expression level

```{r enrichment_analyses, echo=FALSE}
enrichRes <- data.frame("analysis" = NA, "direction" = NA, "geneList" = NA, "nbr_DE" = NA, "nbr_geneList" = NA, 
                        "overlap" = NA, "N" = NA, "pval_enrich" = NA, "pval_deple" = NA, "relative_enrich" = NA,
                        "pval_enrich_random" = NA, "pval_deple_random" = NA, "relative_enrich_random" = NA, 
                        "ctrlExpr" = NA)
enrichRes <- enrichRes[0,]

analysesList <- list(hdbr_sexDE_results_filtered, hdbr_pseudotimeDE_results_filtered, gtex_sexDE_results_filtered, gtex_pseudotimeDE_results_filtered, bayesResults_merged_filtered, bayesResults_merged_filtered)
analysesNameList <- c("prenatal_sexDE", "prenatal_pseudotimeDE", "adult_sexDE", "adult_pseudotimeDE", "prenatal_specific", "shared")
analysesTypeList <- c("sexDE", "pseudotimeDE", "sexDE", "pseudotimeDE", "sexDE", "sexDE")
directions <- c("both", "up", "down")
ctrlExprList <- c(FALSE, TRUE)
for (i in 1:length(analysesList)){
  analyseData <- analysesList[[i]]
  analyseName <- analysesNameList[i]
  analyseType <- analysesTypeList[i]
  print(paste0(">>>>> ", analyseName))
  for (ctrlExpr in ctrlExprList){
    print(paste0(">>> ctrlExpr=", as.character(ctrlExpr)))
    for (myDirection in directions){
      print(paste0("> direction=", myDirection))
      bayes <- FALSE
      bayesDataType <- NULL
      if (analyseName %in% c("prenatal_specific", "shared")){
        bayes <- TRUE
        bayesDataType <- analyseName
      }
      resTmp <- enrichment_analysis_permut(DEresults = analyseData, geneList = downstreamer_genesets_filtered, 
                                           analysisName = analyseName, geneListNames = names(downstreamer_genesets_filtered),
                                           nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection, 
                                           bayes = bayes, bayesDataType = bayesDataType, ctrlExpr = ctrlExpr, geneName = FALSE)
      if (analyseType == "sexDE"){
        if (myDirection == "up"){
          resTmp$direction <- "male"
        } else if (myDirection == "down"){
          resTmp$direction <- "female"
        }
      }
      # add results to the global enrichment abject
      enrichRes <- rbind(enrichRes, resTmp)
    }
  }
}
```


# Write results

Add percentage of the DE genes are in gene list.

```{r add_percent, echo=FALSE}
enrichRes$percent <- enrichRes$overlap/enrichRes$nbr_DE*100
```

Write the complete enrichment results into a file.

```{r write_res, echo=FALSE}
write.table(enrichRes, "data/5-postprocessing/6-Network/results_hyper/enrichment_results.txt", 
            quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
```


# Plot the results

What is the best layout to plot all these information?
We do 2 different plots: one for the bayes results and another for the other results
For both plots, we plot the phenotypes on the y axis and separate the adult and prenatal results on the x axis as well as the sex-DE and pseudotime-DE for each dataset. For each combination, we plot both the relative enrichment value for the geneset of interest as well as a violinplot of the 1000 relative enrichment value for the random genesets.

## Plot for sex-DE and pseudotime-DE results without controlling for gene level of expression

We plot the results for all phenotypes with both direction (up/down or male/female) in the pseudotime-DE and sex-DE analyses for the adult and the prenatal dataset.

First, we format a bit more the data. The chosen threshold to consider a result as significant is 0.05/6 datasets * 2 directions * 8 phenotypes (~96 analyses) = 0.0005.

```{r format4plot, echo=FALSE}
enrichRes_formated <- enrichRes[which(enrichRes$analysis %in% c("prenatal_sexDE", "prenatal_pseudotimeDE", "adult_sexDE", "adult_pseudotimeDE") & enrichRes$direction != "both" & !enrichRes$ctrlExpr), ]
# create groups for facet
enrichRes_formated$dataset <- ifelse(enrichRes_formated$analysis %in% c("prenatal_sexDE", "prenatal_pseudotimeDE"), "prenatal", "adult")
# add * for signif enrichment or depletion (with a threshold of 0.05/96=0.0005)
enrichRes_formated$signif <- ifelse(enrichRes_formated$relative_enrich > 1, 
                                    ifelse(enrichRes_formated$pval_enrich<0.0005 & enrichRes_formated$permut_pval < 0.0005, "*", ""),
                                    ifelse(enrichRes_formated$pval_deple<0.0005 & enrichRes_formated$permut_pval < 0.0005, "*", ""))
# order the different direction of effect
enrichRes_formated$direction <- factor(enrichRes_formated$direction,
                                       levels = c("female", "male", "down", "up"))
# split network names into disease and metaBrain Dataset
enrichRes_formated2 <- separate_wider_delim(data = enrichRes_formated, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_formated2$metaBrainData <- factor(enrichRes_formated2$metaBrainData, 
                                            levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                            labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_formated2$disease <- as.factor(enrichRes_formated2$disease)
levels(enrichRes_formated2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                    "Multiple sclerosis" = "MS",
                                    "Schizophrenia" = "Schizophrenia")
# rename the 4 analyses
enrichRes_formated2$analysis <- factor(enrichRes_formated2$analysis,
                                       levels = c("adult_pseudotimeDE", "adult_sexDE", "prenatal_pseudotimeDE", "prenatal_sexDE"),
                                       labels = c("Adult\nPseudotime-DE", "Adult\nSex-DE", "Prenatal\nPseudotime-DE", "Prenatal\nSex-DE"))

# extract and transform the results for the random set of genes
enrichRes_random_formated2 <- enrichRes_formated2[, c("disease", "metaBrainData", "analysis", "direction", "dataset", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated2$pval_enrich_random <- as.numeric(enrichRes_random_formated2$pval_enrich_random)
enrichRes_random_formated2$pval_deple_random <- as.numeric(enrichRes_random_formated2$pval_deple_random)
enrichRes_random_formated2$relative_enrich_random <- as.numeric(enrichRes_random_formated2$relative_enrich_random)
```


```{r plot, echo=FALSE}
p <- ggplot(enrichRes_random_formated2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "darkred", "male" = "darkblue", "down" = "chocolate", "up" = "aquamarine4"))
p <- p + geom_point(data = enrichRes_formated2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated2, aes(y = disease, x = ifelse(relative_enrich>1, relative_enrich + 0.5, relative_enrich - 0.5), label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue", "down" = "chocolate", "up" = "aquamarine4"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none")
p <- p + coord_cartesian(xlim = c(0,4))
p
# save plot
ggsave("enrichment_plot_sexDE_pseudotimeDE_adult_prenatal.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 20, height = 12)
```

## Plot for bayes results without controlling for gene level of expression

We plot the results for all phenotypes with both direction (male/female) in the prenatal-specific and shared sex-DE genes.

First, we format a bit more the data. The chosen threshold to consider a result as significant is 0.05/6 datasets * 2 directions * 12 phenotypes (~144 analyses) = 0.0003.

```{r format4plot_bayes, echo=FALSE}
enrichRes_formated_bayes <- enrichRes[which(enrichRes$analysis %in% c("prenatal_specific", "shared") & enrichRes$direction != "both" & !enrichRes$ctrlExpr), ]
# add * for signif enrichment or depletion (with a threshold of 0.05/96=0.0005)
enrichRes_formated_bayes$signif <- ifelse(enrichRes_formated_bayes$relative_enrich > 1, 
                                    ifelse(enrichRes_formated_bayes$pval_enrich<0.0005 & enrichRes_formated_bayes$permut_pval < 0.0005, "*", ""),
                                    ifelse(enrichRes_formated_bayes$pval_deple<0.0005 & enrichRes_formated_bayes$permut_pval < 0.0005, "*", ""))
# order the different direction of effect
enrichRes_formated_bayes$direction <- factor(enrichRes_formated_bayes$direction,
                                       levels = c("female", "male"))
# split network names into disease and metaBrain Dataset
enrichRes_formated_bayes2 <- separate_wider_delim(data = enrichRes_formated_bayes, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_formated_bayes2$metaBrainData <- factor(enrichRes_formated_bayes2$metaBrainData, 
                                                  levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                                  labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_formated_bayes2$disease <- as.factor(enrichRes_formated_bayes2$disease)
levels(enrichRes_formated_bayes2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                                  "Multiple sclerosis" = "MS",
                                                  "Schizophrenia" = "Schizophrenia")
# rename the 4 analyses
enrichRes_formated_bayes2$analysis <- factor(enrichRes_formated_bayes2$analysis,
                                             levels = c("prenatal_specific", "shared"),
                                             labels = c("Prenatal\nspecific", "Shared"))

# extract and transform the results for the random set of genes
enrichRes_random_formated_bayes2 <- enrichRes_formated_bayes2[, c("disease", "metaBrainData", "analysis", "direction", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated_bayes2$pval_enrich_random <- as.numeric(enrichRes_random_formated_bayes2$pval_enrich_random)
enrichRes_random_formated_bayes2$pval_deple_random <- as.numeric(enrichRes_random_formated_bayes2$pval_deple_random)
enrichRes_random_formated_bayes2$relative_enrich_random <- as.numeric(enrichRes_random_formated_bayes2$relative_enrich_random)
```


```{r plot_bayes, echo=FALSE}
p <- ggplot(enrichRes_random_formated_bayes2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + geom_point(data = enrichRes_formated_bayes2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated_bayes2, aes(y = disease, x = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.2), label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none")
p <- p + coord_cartesian(xlim = c(0,3.5))
p
# save plot
ggsave("enrichment_plot_sexDE_bayes.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 15, height = 12)
```


## Plot for sex-DE and pseudotime-DE results controlling for gene level of expression

We plot the results for all phenotypes with both direction (up/down or male/female) in the pseudotime-DE and sex-DE analyses for the adult and the prenatal dataset.

First, we format a bit more the data. The chosen threshold to consider a result as significant is 0.05/6 datasets * 2 directions * 8 phenotypes (~96 analyses) = 0.0005.

```{r format4plot_ctrlExpr, echo=FALSE}
enrichRes_formated_ctrlExpr <- enrichRes[which(enrichRes$analysis %in% c("prenatal_sexDE", "prenatal_pseudotimeDE", "adult_sexDE", "adult_pseudotimeDE") & enrichRes$direction != "both" & enrichRes$ctrlExpr), ]
# create groups for facet
enrichRes_formated_ctrlExpr$dataset <- ifelse(enrichRes_formated_ctrlExpr$analysis %in% c("prenatal_sexDE", "prenatal_pseudotimeDE"), "prenatal", "adult")
# add * for signif enrichment or depletion (with a threshold of 0.05/96=0.0005)
enrichRes_formated_ctrlExpr$signif <- ifelse(enrichRes_formated_ctrlExpr$relative_enrich > 1, 
                                             ifelse(enrichRes_formated_ctrlExpr$pval_enrich<0.0005 & enrichRes_formated_ctrlExpr$permut_pval < 0.0005, "*", ""),
                                             ifelse(enrichRes_formated_ctrlExpr$pval_deple<0.0005 & enrichRes_formated_ctrlExpr$permut_pval < 0.0005, "*", ""))
# order the different direction of effect
enrichRes_formated_ctrlExpr$direction <- factor(enrichRes_formated_ctrlExpr$direction,
                                                levels = c("female", "male", "down", "up"))
# split network names into disease and metaBrain Dataset
enrichRes_formated_ctrlExpr2 <- separate_wider_delim(data = enrichRes_formated_ctrlExpr, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_formated_ctrlExpr2$metaBrainData <- factor(enrichRes_formated_ctrlExpr2$metaBrainData, 
                                            levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                            labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_formated_ctrlExpr2$disease <- as.factor(enrichRes_formated_ctrlExpr2$disease)
levels(enrichRes_formated_ctrlExpr2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                    "Multiple sclerosis" = "MS",
                                    "Schizophrenia" = "Schizophrenia")
# rename the 4 analyses
enrichRes_formated_ctrlExpr2$analysis <- factor(enrichRes_formated_ctrlExpr2$analysis,
                                       levels = c("adult_pseudotimeDE", "adult_sexDE", "prenatal_pseudotimeDE", "prenatal_sexDE"),
                                       labels = c("Adult\nPseudotime-DE", "Adult\nSex-DE", "Prenatal\nPseudotime-DE", "Prenatal\nSex-DE"))

# extract and transform the results for the random set of genes
enrichRes_random_formated_ctrlExpr2 <- enrichRes_formated_ctrlExpr2[, c("disease", "metaBrainData", "analysis", "direction", "dataset", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated_ctrlExpr2$pval_enrich_random <- as.numeric(enrichRes_random_formated_ctrlExpr2$pval_enrich_random)
enrichRes_random_formated_ctrlExpr2$pval_deple_random <- as.numeric(enrichRes_random_formated_ctrlExpr2$pval_deple_random)
enrichRes_random_formated_ctrlExpr2$relative_enrich_random <- as.numeric(enrichRes_random_formated_ctrlExpr2$relative_enrich_random)
```


```{r plot_ctrlExpr, echo=FALSE}
p <- ggplot(enrichRes_random_formated_ctrlExpr2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "darkred", "male" = "darkblue", "down" = "chocolate", "up" = "aquamarine4"))
p <- p + geom_point(data = enrichRes_formated_ctrlExpr2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated_ctrlExpr2, aes(y = disease, x = ifelse(relative_enrich>1, relative_enrich + 0.5, relative_enrich - 0.5), label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue", "down" = "chocolate", "up" = "aquamarine4"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none")
p <- p + coord_cartesian(xlim = c(0,4))
p
# save plot
ggsave("enrichment_plot_sexDE_pseudotimeDE_adult_prenatal_ctrlExpr.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 20, height = 12)
```

## Plot for bayes results controlling for gene level of expression

We plot the results for all phenotypes with both direction (male/female) in the prenatal-specific and shared sex-DE genes.

First, we format a bit more the data. The chosen threshold to consider a result as significant is 0.05/6 datasets * 2 directions * 12 phenotypes (~144 analyses) = 0.0003.

```{r format4plot_bayes_ctrlExpr, echo=FALSE}
enrichRes_formated_bayes_ctrlExpr <- enrichRes[which(enrichRes$analysis %in% c("prenatal_specific", "shared") & enrichRes$direction != "both" & enrichRes$ctrlExpr), ]
# add * for signif enrichment or depletion (with a threshold of 0.05/96=0.0005)
enrichRes_formated_bayes_ctrlExpr$signif <- ifelse(enrichRes_formated_bayes_ctrlExpr$relative_enrich > 1, 
                                    ifelse(enrichRes_formated_bayes_ctrlExpr$pval_enrich<0.0005 & enrichRes_formated_bayes_ctrlExpr$permut_pval < 0.0005, "*", ""),
                                    ifelse(enrichRes_formated_bayes_ctrlExpr$pval_deple<0.0005 & enrichRes_formated_bayes_ctrlExpr$permut_pval < 0.0005, "*", ""))
# order the different direction of effect
enrichRes_formated_bayes_ctrlExpr$direction <- factor(enrichRes_formated_bayes_ctrlExpr$direction,
                                       levels = c("female", "male"))
# split network names into disease and metaBrain Dataset
enrichRes_formated_bayes_ctrlExpr2 <- separate_wider_delim(data = enrichRes_formated_bayes_ctrlExpr, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_formated_bayes_ctrlExpr2$metaBrainData <- factor(enrichRes_formated_bayes_ctrlExpr2$metaBrainData, 
                                                  levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                                  labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_formated_bayes_ctrlExpr2$disease <- as.factor(enrichRes_formated_bayes_ctrlExpr2$disease)
levels(enrichRes_formated_bayes_ctrlExpr2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                                  "Multiple sclerosis" = "MS",
                                                  "Schizophrenia" = "Schizophrenia")
# rename the 4 analyses
enrichRes_formated_bayes_ctrlExpr2$analysis <- factor(enrichRes_formated_bayes_ctrlExpr2$analysis,
                                             levels = c("prenatal_specific", "shared"),
                                             labels = c("Prenatal\nspecific", "Shared"))

# extract and transform the results for the random set of genes
enrichRes_random_formated_bayes_ctrlExpr2 <- enrichRes_formated_bayes_ctrlExpr2[, c("disease", "metaBrainData", "analysis", "direction", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated_bayes_ctrlExpr2$pval_enrich_random <- as.numeric(enrichRes_random_formated_bayes_ctrlExpr2$pval_enrich_random)
enrichRes_random_formated_bayes_ctrlExpr2$pval_deple_random <- as.numeric(enrichRes_random_formated_bayes_ctrlExpr2$pval_deple_random)
enrichRes_random_formated_bayes_ctrlExpr2$relative_enrich_random <- as.numeric(enrichRes_random_formated_bayes_ctrlExpr2$relative_enrich_random)
```


```{r plot_bayes_ctrlExpr, echo=FALSE}
p <- ggplot(enrichRes_random_formated_bayes_ctrlExpr2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + geom_point(data = enrichRes_formated_bayes_ctrlExpr2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated_bayes_ctrlExpr2, aes(y = disease, x = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.2), label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none")
p <- p + coord_cartesian(xlim = c(0,3.5))
p
# save plot
ggsave("enrichment_plot_sexDE_bayes_ctrlExpr.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 15, height = 12)
```

## Plot for supplementary figure

### Plot enrichment results for sex-DE results controlling for gene expression

First, we need to select the data we want to plot and format it.

```{r select4plot_sexDE_ctrlExpr, echo=FALSE}
# select analyses of interest
enrichRes_sexDE_formated_ctrlExpr <- enrichRes[which(enrichRes$analysis %in% c("prenatal_sexDE", "adult_sexDE") &
                                                       enrichRes$direction != "both" & enrichRes$ctrlExpr), ]
```

```{r format4plot_sexDE_ctrlExpr, echo=FALSE}
# add * for signif enrichment or depletion (with a threshold of 0.05/8*8=0.00078)
enrichRes_sexDE_formated_ctrlExpr$signif <- ifelse(enrichRes_sexDE_formated_ctrlExpr$pval_enrich<0.00078 & 
                                                     enrichRes_sexDE_formated_ctrlExpr$permut_pval < 0.0005, "*", "")
# order the different direction of effect
enrichRes_sexDE_formated_ctrlExpr$direction <- factor(enrichRes_sexDE_formated_ctrlExpr$direction,
                                                      levels = c("female", "male"))
# split network names into disease and metaBrain Dataset
enrichRes_sexDE_formated_ctrlExpr2 <- separate_wider_delim(data = enrichRes_sexDE_formated_ctrlExpr, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_sexDE_formated_ctrlExpr2$metaBrainData <- factor(enrichRes_sexDE_formated_ctrlExpr2$metaBrainData, 
                                                           levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                                           labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_sexDE_formated_ctrlExpr2$disease <- as.factor(enrichRes_sexDE_formated_ctrlExpr2$disease)
levels(enrichRes_sexDE_formated_ctrlExpr2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                                           "Multiple sclerosis" = "MS",
                                                           "Schizophrenia" = "Schizophrenia")
# rename the 2 analyses
enrichRes_sexDE_formated_ctrlExpr2$analysis <- factor(enrichRes_sexDE_formated_ctrlExpr2$analysis,
                                                      levels = c("adult_sexDE", "prenatal_sexDE"),
                                                      labels = c("Adult", "Prenatal"))

# extract and transform the results for the random set of genes
enrichRes_sexDE_random_formated_ctrlExpr2 <- enrichRes_sexDE_formated_ctrlExpr2[, c("disease", "metaBrainData", "analysis", "direction", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_sexDE_random_formated_ctrlExpr2$pval_enrich_random <- as.numeric(enrichRes_sexDE_random_formated_ctrlExpr2$pval_enrich_random)
enrichRes_sexDE_random_formated_ctrlExpr2$pval_deple_random <- as.numeric(enrichRes_sexDE_random_formated_ctrlExpr2$pval_deple_random)
enrichRes_sexDE_random_formated_ctrlExpr2$relative_enrich_random <- as.numeric(enrichRes_sexDE_random_formated_ctrlExpr2$relative_enrich_random)
```

Then, we plot the enrichment results.

```{r plot_sexDE_ctrlExpr, echo=FALSE}
p <- ggplot(enrichRes_sexDE_random_formated_ctrlExpr2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "#ae4d4d", "male" = "#4d4dae"))
p <- p + geom_point(data = enrichRes_sexDE_formated_ctrlExpr2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_sexDE_formated_ctrlExpr2, aes(y = disease, x = relative_enrich + 0.3, label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none", color="none", fill="none")
p <- p + coord_cartesian(xlim = c(0,4))
p
# save plot
ggsave("enrichment_plot_sexDE_adult_prenatal_diseaseOnly_ctrlExpr.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 10, height = 10)
```


### Plot enrichment results for sex-DE results controlling for gene expression

First, we need tot select the data we want to plot and format it.

```{r select4plot_bayes_ctrlExpr, echo=FALSE}
# select analyses of interest
enrichRes_bayes_formated_ctrlExpr <- enrichRes[which(enrichRes$analysis %in% c("prenatal_specific", "shared") &
                                                       enrichRes$direction != "both" & enrichRes$ctrlExpr), ]
```

```{r format4plot_bayes_ctrlExpr, echo=FALSE}
# add * for signif enrichment or depletion (with a threshold of 0.05/8*8=0.00078)
enrichRes_bayes_formated_ctrlExpr$signif <- ifelse(enrichRes_bayes_formated_ctrlExpr$pval_enrich<0.00078 & 
                                                     enrichRes_bayes_formated_ctrlExpr$permut_pval < 0.05, "*", "")
# order the different direction of effect
enrichRes_bayes_formated_ctrlExpr$direction <- factor(enrichRes_bayes_formated_ctrlExpr$direction,
                                                      levels = c("female", "male"))
# split network names into disease and metaBrain Dataset
enrichRes_bayes_formated_ctrlExpr2 <- separate_wider_delim(data = enrichRes_bayes_formated_ctrlExpr, cols = "geneList", 
                                                           delim = ".", names = c("disease", "metaBrainData")) 
# rename metaBrainData into shorter version
enrichRes_bayes_formated_ctrlExpr2$metaBrainData <- factor(enrichRes_bayes_formated_ctrlExpr2$metaBrainData, 
                                                           levels = c("MetaBrain", "MetaBrainCerebellum", "MetaBrainCortex"), 
                                                           labels = c("MetaBrain" = "AllBrain", "MetaBrainCerebellum" = "Cerebellum", "MetaBrainCortex" = "Cortex"))
# order diseases
enrichRes_bayes_formated_ctrlExpr2$disease <- as.factor(enrichRes_bayes_formated_ctrlExpr2$disease)
levels(enrichRes_bayes_formated_ctrlExpr2$disease) <- list("Amyotrophic lateral\nsclerosis" = "ALS",
                                                           "Multiple sclerosis" = "MS",
                                                           "Schizophrenia" = "Schizophrenia")
# rename the 2 analyses
enrichRes_bayes_formated_ctrlExpr2$analysis <- factor(enrichRes_bayes_formated_ctrlExpr2$analysis,
                                                      levels = c("prenatal_specific", "shared"),
                                                      labels = c("Prenatal\nspecific", "Shared"))

# extract and transform the results for the random set of genes
enrichRes_bayes_random_formated_ctrlExpr2 <- enrichRes_bayes_formated_ctrlExpr2[, c("disease", "metaBrainData", "analysis", "direction", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_bayes_random_formated_ctrlExpr2$pval_enrich_random <- as.numeric(enrichRes_bayes_random_formated_ctrlExpr2$pval_enrich_random)
enrichRes_bayes_random_formated_ctrlExpr2$pval_deple_random <- as.numeric(enrichRes_bayes_random_formated_ctrlExpr2$pval_deple_random)
enrichRes_bayes_random_formated_ctrlExpr2$relative_enrich_random <- as.numeric(enrichRes_bayes_random_formated_ctrlExpr2$relative_enrich_random)
```


```{r plot_bayes_ctrlExpr, echo=FALSE}
p <- ggplot(enrichRes_bayes_random_formated_ctrlExpr2, aes(y = disease, x = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin(linewidth = 0.1)
p <- p + geom_vline(xintercept = 1)
p <- p + facet_grid(cols = vars(analysis), rows = vars(metaBrainData), scales = "free_y", space = "free_y", drop = TRUE)
p <- p + scale_fill_manual(values = c("female" = "#ae4d4d", "male" = "#4d4dae"))
p <- p + geom_point(data = enrichRes_bayes_formated_ctrlExpr2, aes(y = disease, x = relative_enrich, color = direction), 
                    position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_bayes_formated_ctrlExpr2, aes(y = disease, x = relative_enrich + 0.2, label = signif),
                   position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "darkred", "male" = "darkblue"))
p <- p + theme_minimal()
p <- p + ylab("") + xlab("Relative enrichment")
p <- p + guides(alpha="none", color="none", fill="none")
p <- p + coord_cartesian(xlim = c(0,3.5))
p
# save plot
ggsave("enrichment_plot_sexDE_bayes_diseaseOnly_ctrlExpr.png",
      path = "images/5-postprocessing/6-Network/network_hyper/",
      units = "cm", width = 10, height = 10)
```

### Save data for supplementary table

Save the data from both plots for supplementary table.

```{r save_data, echo=FALSE}
# select the data of interest
enrichRes_formated_ctrlExpr_toSave <- enrichRes[which(enrichRes$analysis %in% c("prenatal_sexDE", "adult_sexDE", "prenatal_specific", "shared") &
                                                        enrichRes$direction != "both" & enrichRes$ctrlExpr), ]
# split network names into disease and metaBrain Dataset
enrichRes_formated_ctrlExpr_toSave <- separate_wider_delim(data = enrichRes_formated_ctrlExpr_toSave, cols = "geneList", delim = ".", names = c("disease", "metaBrainData")) 
# format the data by selecting the columns of interest
enrichRes_formated_ctrlExpr_toSave <- enrichRes_formated_ctrlExpr_toSave[, c("analysis", "direction", "disease", "metaBrainData", "nbr_DE", "nbr_geneList", "overlap", "N",  "pval_enrich", "relative_enrich", "pval_enrich_random", "relative_enrich_random", "permut_pval")]
# format the analysis column
enrichRes_formated_ctrlExpr_toSave$analysis <- str_remove(enrichRes_formated_ctrlExpr_toSave$analysis, "_sexDE")
# save in file
write.table(enrichRes_formated_ctrlExpr_toSave, file = "data/5-postprocessing/6-Network/results_hyper/enrichment_results_formated.txt", 
            quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
```



# Session info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

